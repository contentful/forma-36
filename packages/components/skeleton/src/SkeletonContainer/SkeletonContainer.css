@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.SkeletonContainer {
  display: block;
}
