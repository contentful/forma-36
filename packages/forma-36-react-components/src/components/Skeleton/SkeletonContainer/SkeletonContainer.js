import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './SkeletonContainer.css';

export class SkeletonContainer extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    ariaLabel: PropTypes.string,
    width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    preserveAspectRatio: PropTypes.string,
    clipId: PropTypes.string,
    gradientId: PropTypes.string,
    backgroundColor: PropTypes.string,
    backgroundOpacity: PropTypes.number,
    animate: PropTypes.bool,
    speed: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    foregroundColor: PropTypes.string,
    foregroundOpacity: PropTypes.number,
    svgWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    svgHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  };

  static defaultProps = {
    extraClassNames: undefined,
    testId: 'cf-ui-skeleton-form',
    ariaLabel: 'Loading component...',
    width: '100%',
    height: '100%',
    preserveAspectRatio: 'xMidYMid meet',
    clipId: 'cf-ui-skeleton-clip-id',
    gradientId: 'cf-ui-skeleton-clip-gradient',
    backgroundColor: '#e5ebed',
    backgroundOpacity: 1,
    animate: true,
    speed: 2,
    foregroundColor: '#f7f9fa',
    foregroundOpacity: 1,
    svgWidth: '100%',
    svgHeight: '100%',
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      ariaLabel,
      width,
      height,
      preserveAspectRatio,
      clipId,
      gradientId,
      backgroundColor,
      backgroundOpacity,
      animate,
      speed,
      foregroundColor,
      foregroundOpacity,
      svgWidth,
      svgHeight,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.SkeletonContainer, extraClassNames);

    return (
      <svg
        role="img"
        className={classNames}
        aria-label={ariaLabel}
        preserveAspectRatio={preserveAspectRatio}
        width={svgWidth}
        height={svgHeight}
        {...otherProps}
      >
        {ariaLabel ? <title>{ariaLabel}</title> : null}
        <rect
          x="0"
          y="0"
          width={width}
          height={height}
          clipPath={`url(#${clipId})`}
          style={{ fill: `url(#${gradientId})` }}
        />

        <defs>
          <clipPath id={clipId}>{children}</clipPath>

          <linearGradient id={gradientId}>
            <stop
              offset="0%"
              stopColor={backgroundColor}
              stopOpacity={backgroundOpacity}
            >
              {animate && (
                <animate
                  id="animation1"
                  attributeName="stop-color"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
            <stop
              offset="50%"
              stopColor={foregroundColor}
              stopOpacity={foregroundOpacity}
            >
              {animate && (
                <animate
                  attributeName="stop-color"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  begin="animation1.begin+0.25s"
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
            <stop
              offset="100%"
              stopColor={backgroundColor}
              stopOpacity={backgroundOpacity}
            >
              {animate && (
                <animate
                  attributeName="stop-color"
                  begin="animation1.begin+0.5s"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
          </linearGradient>
        </defs>
      </svg>
    );
  }
}

export default SkeletonContainer;
