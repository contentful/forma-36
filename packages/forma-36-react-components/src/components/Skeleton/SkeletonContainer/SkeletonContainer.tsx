import React from 'react';
import cn from 'classnames';
import styles from './SkeletonContainer.css';

export interface SkeletonContainerProps {
  width?: number | string;
  height?: number | string;
  preserveAspectRatio?: string;
  clipId?: string;
  gradientId?: string;
  animateId?: string;
  backgroundColor?: string;
  backgroundOpacity?: number;
  animate?: boolean;
  speed?: number | string;
  foregroundColor?: string;
  foregroundOpacity?: number;
  svgWidth?: string | number;
  svgHeight?: string | number;

  ariaLabel?: string;
  className?: string;
  testId?: string;
  children: React.ReactNode;
}

let idCounter = 0;

export function SkeletonContainer({
  className,
  children,
  testId = 'cf-ui-skeleton-form',
  ariaLabel = 'Loading component...',
  width = '100%',
  height = '100%',
  preserveAspectRatio = 'xMidYMid meet',
  clipId = `cf-ui-skeleton-clip-${idCounter++}`,
  gradientId = `cf-ui-skeleton-clip-gradient-${idCounter++}`,
  animateId = `animation-${idCounter++}`,
  backgroundColor = '#e5ebed',
  backgroundOpacity = 1,
  animate = true,
  speed = 2,
  foregroundColor = '#f7f9fa',
  foregroundOpacity = 1,
  svgWidth = '100%',
  svgHeight = '100%',
  ...otherProps
}: SkeletonContainerProps): React.ReactElement {
  const classNames = cn(styles['SkeletonContainer'], className);

  return (
    <svg
      role="img"
      className={classNames}
      aria-label={ariaLabel}
      preserveAspectRatio={preserveAspectRatio}
      width={svgWidth}
      height={svgHeight}
      data-test-id={testId}
      {...otherProps}
    >
      {ariaLabel ? <title>{ariaLabel}</title> : null}
      <rect
        x="0"
        y="0"
        width={width}
        height={height}
        clipPath={`url(#${clipId})`}
        style={{ fill: `url(#${gradientId})` }}
      />

      <defs>
        <clipPath id={clipId}>{children}</clipPath>

        <linearGradient id={gradientId}>
          <stop
            offset="0%"
            stopColor={backgroundColor}
            stopOpacity={backgroundOpacity}
          >
            {animate && (
              <animate
                id={animateId}
                attributeName="stop-color"
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
          <stop
            offset="50%"
            stopColor={foregroundColor}
            stopOpacity={foregroundOpacity}
          >
            {animate && (
              <animate
                attributeName="stop-color"
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                begin={`${animateId}.begin+0.25s`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
          <stop
            offset="100%"
            stopColor={backgroundColor}
            stopOpacity={backgroundOpacity}
          >
            {animate && (
              <animate
                attributeName="stop-color"
                begin={`${animateId}.begin+0.5s`}
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
        </linearGradient>
      </defs>
    </svg>
  );
}
