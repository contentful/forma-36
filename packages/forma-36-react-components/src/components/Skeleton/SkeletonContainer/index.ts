export * from './SkeletonContainer';
export { default } from './SkeletonContainer';
