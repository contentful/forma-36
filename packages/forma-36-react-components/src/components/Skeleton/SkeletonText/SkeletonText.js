import React from 'react';
import PropTypes from 'prop-types';

export const SkeletonTextPropTypes = {
  numberOfLines: PropTypes.number,
  width: PropTypes.number,
  offsetTop: PropTypes.number,
  offsetLeft: PropTypes.number,
  lineHeight: PropTypes.number,
  marginBottom: PropTypes.number,
};

export class SkeletonText extends React.Component {
  static propTypes = SkeletonTextPropTypes;

  static defaultProps = {
    numberOfLines: 1,
    width: undefined,
    offsetTop: 0,
    offsetLeft: 0,
    lineHeight: 21,
    marginBottom: 20,
  };

  getLineWidth = lastLine => {
    if (this.props.width) {
      return this.props.width;
    }
    return lastLine ? '80%' : '100%';
  };

  render() {
    const {
      numberOfLines,
      offsetLeft,
      offsetTop,
      lineHeight,
      marginBottom,
    } = this.props;

    return (
      <React.Fragment>
        {Array.from(Array(numberOfLines)).map((_, index) => (
          <rect
            key={`skeleton-display-text-${index}`} // eslint-disable-line
            x={offsetLeft}
            y={index * (lineHeight + marginBottom) + offsetTop}
            rx="0"
            ry="0"
            width={this.getLineWidth(
              numberOfLines > 1 && numberOfLines - index === 1,
            )}
            height={lineHeight}
          />
        ))}
      </React.Fragment>
    );
  }
}

export default SkeletonText;
