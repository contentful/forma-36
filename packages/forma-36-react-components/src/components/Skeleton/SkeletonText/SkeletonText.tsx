import React, { Component } from 'react';

type stringOrNumber = string | number;

export interface SkeletonTextProps {
  numberOfLines?: number;
  offsetTop?: stringOrNumber;
  offsetLeft?: stringOrNumber;
  lineHeight?: stringOrNumber;
  marginBottom?: stringOrNumber;
  width?: stringOrNumber;
}

const defaultProps: Partial<SkeletonTextProps> = {
  numberOfLines: 1,
  offsetTop: 0,
  offsetLeft: 0,
  lineHeight: 21,
  marginBottom: 20,
};

export class SkeletonText extends Component<SkeletonTextProps> {
  static defaultProps = defaultProps;

  getLineWidth = (lastLine: boolean) => {
    if (this.props.width) {
      return this.props.width;
    }
    return lastLine ? '80%' : '100%';
  };

  render() {
    const {
      numberOfLines,
      offsetLeft,
      offsetTop,
      lineHeight,
      marginBottom,
    } = this.props;

    return (
      <React.Fragment>
        {Array.from(Array(numberOfLines)).map((_, index) => (
          <rect
            key={`skeleton-display-text-${index}`} // eslint-disable-line
            x={offsetLeft}
            y={
              index * (+lineHeight! + +marginBottom!) + +offsetTop! // eslint-disable-line @typescript-eslint/no-non-null-assertion
            }
            rx="0"
            ry="0"
            width={this.getLineWidth(
              numberOfLines! > 1 && numberOfLines! - index === 1, // eslint-disable-line @typescript-eslint/no-non-null-assertion
            )}
            height={lineHeight}
          />
        ))}
      </React.Fragment>
    );
  }
}

export default SkeletonText;
