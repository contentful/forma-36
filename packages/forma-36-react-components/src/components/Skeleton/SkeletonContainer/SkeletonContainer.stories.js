import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean, number } from '@storybook/addon-knobs';

import SkeletonBodyText from '../SkeletonBodyText';
import SkeletonDisplayText from '../SkeletonDisplayText';
import SkeletonContainer from './SkeletonContainer';

storiesOf('Components|Skeleton/SkeletonContainer', module).add(
  'default',
  () => (
    <SkeletonContainer
      extraClassNames={text('Extra Class Names', '')}
      animate={boolean('Animate', true)}
      width={text('Width', '100%')}
      height={text('Height', '100')}
      backgroundColor={text('Primary color', '#e5ebed')}
      foregroundColor={text('Secondary color', '#f7f9fa')}
      speed={number('Speed', 2)}
    >
      <SkeletonDisplayText numberOfLines={1} />
      <SkeletonBodyText numberOfLines={3} offsetTop={35} />
    </SkeletonContainer>
  ),
);
