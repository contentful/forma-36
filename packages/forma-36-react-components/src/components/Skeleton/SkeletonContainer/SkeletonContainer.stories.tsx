import React from 'react';

import SkeletonBodyText from '../SkeletonBodyText';
import SkeletonDisplayText from '../SkeletonDisplayText';
import SkeletonContainer, { SkeletonContainerProps } from './SkeletonContainer';

export default {
  title: 'Components/Skeleton/SkeletonContainer',
  component: SkeletonContainer,
  parameters: {
    propTypes: [SkeletonContainer['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
  },
};

export const Basic = (args: SkeletonContainerProps) => (
  <SkeletonContainer {...args}>
    <SkeletonDisplayText numberOfLines={1} />
    <SkeletonBodyText numberOfLines={3} offsetTop={35} />
  </SkeletonContainer>
);

Basic.args = {
  width: '100%',
  height: '100',
  backgroundColor: '#e5ebed',
  foregroundColor: '#f7f9fa',
  speed: 2,
};
