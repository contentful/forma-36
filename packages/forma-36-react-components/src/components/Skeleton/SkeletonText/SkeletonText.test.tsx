import React from 'react';
import { shallow } from 'enzyme';
import axe from '../../../utils/axeHelper';
import SkeletonContainer from '../SkeletonContainer';
import SkeletonText from './SkeletonText';

it('renders the component', () => {
  const output = shallow(
    <SkeletonContainer>
      <SkeletonText />
    </SkeletonContainer>,
  );

  expect(output).toMatchSnapshot();
});

it('renders 3 lines of skeleton text', () => {
  const output = shallow(
    <SkeletonContainer>
      <SkeletonText numberOfLines={3} />
    </SkeletonContainer>,
  );

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = shallow(
    <SkeletonContainer>
      <SkeletonText />
    </SkeletonContainer>,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
