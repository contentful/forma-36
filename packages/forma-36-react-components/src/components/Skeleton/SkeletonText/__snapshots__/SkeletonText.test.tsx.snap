// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 3 lines of skeleton text 1`] = `
<svg
  aria-label="Loading component..."
  className="SkeletonContainer"
  data-test-id="cf-ui-skeleton-form"
  height="100%"
  preserveAspectRatio="xMidYMid meet"
  role="img"
  width="100%"
>
  <title>
    Loading component...
  </title>
  <rect
    clipPath="url(#cf-ui-skeleton-clip-2)"
    height="100%"
    style={
      Object {
        "fill": "url(#cf-ui-skeleton-clip-gradient-3)",
      }
    }
    width="100%"
    x="0"
    y="0"
  />
  <defs>
    <clipPath
      id="cf-ui-skeleton-clip-2"
    >
      <SkeletonText
        lineHeight={21}
        marginBottom={20}
        numberOfLines={3}
        offsetLeft={0}
        offsetTop={0}
      />
    </clipPath>
    <linearGradient
      id="cf-ui-skeleton-clip-gradient-3"
    >
      <stop
        offset="0%"
        stopColor="#e5ebed"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          dur="2s"
          id="animation1"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="50%"
        stopColor="#f7f9fa"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          begin="animation1.begin+0.25s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="100%"
        stopColor="#e5ebed"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          begin="animation1.begin+0.5s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
    </linearGradient>
  </defs>
</svg>
`;

exports[`renders the component 1`] = `
<svg
  aria-label="Loading component..."
  className="SkeletonContainer"
  data-test-id="cf-ui-skeleton-form"
  height="100%"
  preserveAspectRatio="xMidYMid meet"
  role="img"
  width="100%"
>
  <title>
    Loading component...
  </title>
  <rect
    clipPath="url(#cf-ui-skeleton-clip-0)"
    height="100%"
    style={
      Object {
        "fill": "url(#cf-ui-skeleton-clip-gradient-1)",
      }
    }
    width="100%"
    x="0"
    y="0"
  />
  <defs>
    <clipPath
      id="cf-ui-skeleton-clip-0"
    >
      <SkeletonText
        lineHeight={21}
        marginBottom={20}
        numberOfLines={1}
        offsetLeft={0}
        offsetTop={0}
      />
    </clipPath>
    <linearGradient
      id="cf-ui-skeleton-clip-gradient-1"
    >
      <stop
        offset="0%"
        stopColor="#e5ebed"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          dur="2s"
          id="animation1"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="50%"
        stopColor="#f7f9fa"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          begin="animation1.begin+0.25s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="100%"
        stopColor="#e5ebed"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          begin="animation1.begin+0.5s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
    </linearGradient>
  </defs>
</svg>
`;
